pilot/pkg/proxy/envoy/v2/lds.go

```diff
@@ -65,49 +111,118 @@ func (s *DiscoveryServer) StreamListeners(stream xdsapi.ListenerDiscoveryService
 			if !ok {
 				return receiveError
 			}
+			nt, err := model.ParseServiceNode(discReq.Node.Id)
+			if err != nil {
+				return err
+			}
 			if !initialRequest {
-				log.Debugf("LDS ACK from Envoy for client %q has version %q and Nonce %q for request", discReq.GetVersionInfo(), discReq.GetResponseNonce())
+				log.Debugf("LDS: ACK from Envoy for client %q has version %q and Nonce %q for request", discReq.GetVersionInfo(), discReq.GetResponseNonce())
 				continue
 			}
+			node.ID = discReq.Node.Id
+			node.Type = nt.Type
+
+			nodeID = nt.ID
+			addLdsCon(nodeID, con)
+			log.Infof("LDS: StreamListeners %v %s %s", peerAddr, nt.ID, discReq.String())
+
 		case <-ticker.C:
-			if !initialRequest {
+			if initialRequest {
 				// Ignore ticker events until the very first request is processed.
 				continue
 			}
-		}
-		if initialRequest {
-			initialRequest = false
-			log.Debugf("LDS request from  %q received.", peerAddr)
+		case <-con.pushChannel:
 		}
 
-		nt, err := model.ParseServiceNode(discReq.Node.Id)
-		if err != nil {
-			return err
-		}
-		node := model.Proxy{
-			ID:   discReq.Node.Id,
-			Type: nt.Type,
-		}
-		response, err := ListListenersResponse(s.env, node)
-		log.Info(response.String())
+		response, err := con.ldsDiscoveryResponse(s.env, node)
 		if err != nil {
 			return err
 		}
 		err = stream.Send(response)
 		if err != nil {
 			return err
 		}
-		log.Debugf("\nLDS response from  %q, Response: \n%s\n\n", peerAddr, response.String())
+		initialRequest = false
 	}
 }
 
-// FetchListeners implements the DiscoveryServer interface.
-func (s *DiscoveryServer) FetchListeners(ctx context.Context, in *xdsapi.DiscoveryRequest) (*xdsapi.DiscoveryResponse, error) {
-	log.Info("FetchListeners")
-	node, err := model.ParseServiceNode(in.Node.Id)
+// ldsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func ldsPushAll() {
+	if ldsDebug {
+		log.Infoa("LDS cache reset")
+	}
+	ldsClientsMutex.RLock()
```

Comment:
Just push to the channel directly while holding the read lock. The lock means a writer can't add/remove a client out from under you. We do run the risk of a slow writer blocking additions to the map (since our pushChannel is blocking), I suppose. Not sure if its worth worrying about.

Otherwise, at least allocate an array rather than a map, and size it correctly (that way we just run a linear scan, rather than hopping around a map loading each listener to push to its channel).

```go
tmp := make([]*LdsConnection, 0, len(ldsClients))
for _, v := range ldsClients { tmp = append(tmp, v) }
```

Reply:
Good idea, can we:
- do it in a follow up
- do it for both EDS and LDS ( and CDS if it happens before).

Pushink while we held the lock seems dangerous, it may work but I would avoid the risk even if it
means an extra copy.